src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -94,100 +94,91 @@ public virtual void GetPolicies(JsonSerializerOptions options)
-                    if (NameAsString == null)
-                    {
-                        return;
-                    }
-                }
-                else
+                if (NameAsString == null)
                 {
-                    NameAsString = PropertyInfo.Name;
+                    ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameNull(ParentClassType, this);
                 }
+            }
+            else if (options.PropertyNamingPolicy != null)
+            {
+                NameAsString = options.PropertyNamingPolicy.ConvertName(PropertyInfo.Name);
 
-                // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
-                _name = Encoding.UTF8.GetBytes(NameAsString);
-
-                // Set the compare name.
-                if (options.PropertyNameCaseInsensitive)
+                if (NameAsString == null)
                 {
-                    NameUsedToCompareAsString = NameAsString.ToUpperInvariant();
-                    _nameUsedToCompare = Encoding.UTF8.GetBytes(NameUsedToCompareAsString);
-                }
-                else
-                {
-                    NameUsedToCompareAsString = NameAsString;
-                    _nameUsedToCompare = _name;
+                    ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameNull(ParentClassType, this);
                 }
+            }
+            else
+            {
+                NameAsString = PropertyInfo.Name;
+            }
+
+            // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
+            _name = Encoding.UTF8.GetBytes(NameAsString);
+
+            // Set the compare name.
+            if (options.PropertyNameCaseInsensitive)
+            {
+                NameUsedToCompareAsString = NameAsString.ToUpperInvariant();
+                _nameUsedToCompare = Encoding.UTF8.GetBytes(NameUsedToCompareAsString);
+            }
+            else
+            {
+                NameUsedToCompareAsString = NameAsString;
+                _nameUsedToCompare = _name;
+            }
 
-                // Cache the escaped name.
+            // Cache the escaped name.
 #if true
-                // temporary behavior until the writer can accept escaped string.
-                _escapedName = _name;
+            // temporary behavior until the writer can accept escaped string.
+            _escapedName = _name;
 #else
-                
-                int valueIdx = JsonWriterHelper.NeedsEscaping(_name);
-                if (valueIdx == -1)
-                {
-                    _escapedName = _name;
-                }
-                else
-                {
-                    byte[] pooledName = null;
-                    int length = JsonWriterHelper.GetMaxEscapedLength(_name.Length, valueIdx);
+            int valueIdx = JsonWriterHelper.NeedsEscaping(_name);
```

Comment:
In any case, I don't think the #else code block is adding much value, and we should get rid of it, for now.

Reply:
I'll do this when we convert to use that new feature.